### PR TITLE
fix: refactor pointer creation to protect against improper init

### PR DESCRIPTION
**Pull request by kneekey23**
*Description of changes:* This PR refactors the way pointers are created and initialized in Swift to protect against cases where the pointer is created but uninitialized later causing a seg fault when the c code calls back into the Swift code.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: Source/AwsCommonRuntimeKit/Utilities.swift [hunk 1]
```diff
@@ -57,3 +57,25 @@ extension Bool {
         return self ? 1 : 0
     }
 }
+
+func getMutablePointer<T>(memoryType: T?) -> UnsafeMutablePointer<T>? {
```
Comment (justinboswell) [inline Source/AwsCommonRuntimeKit/Utilities.swift:61]: Is it possible to write this as:
```swift
func fromPointer<T, P>(ptr: T?) -> P<T>? {
   if let ptr = ptr {
        let pointer = UnsafeMutablePointer<T>.allocate(capacity: 1)
        pointer.initialize(to: ptr)
        return P<T>(pointer)
    }
    return nil
}
```

Comment (kneekey23) [reply Source/AwsCommonRuntimeKit/Utilities.swift:61]: tried something similar to this but let me try again

File: Source/AwsCommonRuntimeKit/Utilities.swift [hunk 2]
```diff
@@ -57,3 +57,27 @@ extension Bool {
         return self ? 1 : 0
     }
 }
+
+func fromPointer<T, P: PointerConformance>(ptr: T?) -> P? {
+   if let ptr = ptr {
+        let pointer = UnsafeMutablePointer<T>.allocate(capacity: 1)
+        pointer.initialize(to: ptr)
+       return P(OpaquePointer(pointer))
```
Comment (justinboswell) [inline Source/AwsCommonRuntimeKit/Utilities.swift:65]: whitespace is weird

File: Source/AwsCommonRuntimeKit/Utilities.swift [hunk 3]
```diff
@@ -57,3 +57,27 @@ extension Bool {
         return self ? 1 : 0
     }
 }
+
```
Comment (justinboswell) [inline Source/AwsCommonRuntimeKit/Utilities.swift:60]:
```swift
// Ensure that any UnsafeXXXPointer is ALWAYS initialized to either nil or a value in a single call. Prevents the
// case where you create an UnsafeMutableWhatever and do not call `initialize()` on it, resulting in a non-null but // also invalid pointer
```

File: Source/AwsCommonRuntimeKit/auth/credentials/CRTCredentialsProvider.swift [hunk 4]
```diff
@@ -68,24 +65,19 @@ class WrappedCRTCredentialsProvider {
...
+        let pointer: UnsafeMutablePointer<CRTCredentialsProviderShutdownOptions>? = fromOptionalPointer(ptr: shutDownOptions)
+        let shutDownOptionsC = aws_credentials_provider_shutdown_options(shutdown_callback: { userData in
+            guard let userData = userData else {
+                return
+            }
+            let pointer = userData.assumingMemoryBound(to: CRTCredentialsProviderShutdownOptions.self)
+            defer {pointer.deinitializeAndDeallocate()}
+            pointer.pointee.shutDownCallback()
+
+        }, shutdown_user_data: pointer)
```
Comment (justinboswell) [inline Source/AwsCommonRuntimeKit/auth/credentials/CRTCredentialsProvider.swift:78]: Feels like this section needs a format/lint

File: Source/AwsCommonRuntimeKit/auth/credentials/CRTCredentialsProvider.swift [hunk 5]
```diff
@@ -48,14 +48,11 @@ class WrappedCRTCredentialsProvider {
 
         })
         let shutDownOptions = Self.setUpShutDownOptions(shutDownOptions: shutDownOptions)
-        let intPointer = UnsafeMutablePointer<Int>.allocate(capacity: 1)
-        intPointer.pointee = 1
-        let atomicVar = aws_atomic_var(value: UnsafeMutableRawPointer(intPointer))
+        let intPointer: UnsafeMutableRawPointer = fromPointer(ptr: 1)
```
Comment (justinboswell) [inline Source/AwsCommonRuntimeKit/auth/credentials/CRTCredentialsProvider.swift:51]: I hadn't noticed this code before. Does this imply that Swift doesn't have atomics?!

File: Source/AwsCommonRuntimeKit/io/EventLoopGroup.swift [hunk 6]
```diff
@@ -12,24 +12,18 @@ public final class EventLoopGroup {
...
-            shutDownPtr.initialize(to: shutDownOptions)
-            let options = aws_shutdown_callback_options(shutdown_callback_fn: { (userData) in
-                guard let userdata = userData else {
-                    return
-                }
-                let pointer = userdata.assumingMemoryBound(to: ShutDownCallbackOptions.self)
-                defer { pointer.deinitializeAndDeallocate() }
-                pointer.pointee.shutDownCallback(pointer.pointee.semaphore)
+        let shutDownPtr: UnsafeMutablePointer<ShutDownCallbackOptions>? = fromOptionalPointer(ptr: shutDownOptions)
+        let options = aws_shutdown_callback_options(shutdown_callback_fn: { (userData) in
```
Comment (justinboswell) [inline Source/AwsCommonRuntimeKit/io/EventLoopGroup.swift:16]: I've just realized that we've re-used this exact pattern at least 3 times, we should have a function that produces ShutdownCallbackOptions from a closure.

File: Source/AwsCommonRuntimeKit/io/TlsConnectionOptions.swift [hunk 7]
```diff
@@ -9,10 +9,8 @@ public final class TlsConnectionOptions {
 
 	init(_ context: TlsContext, allocator: Allocator) {
 		self.allocator = allocator
-        let connectionOptionsPtr = UnsafeMutablePointer<aws_tls_connection_options>.allocate(capacity: 1)
-        //initialize pointer to empty connection options struct
-        zeroStruct(connectionOptionsPtr)
-        self.rawValue = connectionOptionsPtr
+
+        self.rawValue = allocatePointer()
```
Comment (justinboswell) [inline Source/AwsCommonRuntimeKit/io/TlsConnectionOptions.swift:13]: Formatting?

File: Source/AwsCommonRuntimeKit/io/retryer/CRTAWSRetryToken.swift [hunk 8]
```diff
@@ -7,17 +7,14 @@ public final class CRTAWSRetryToken {
     var rawValue: UnsafeMutablePointer<aws_retry_token>
 
     public init(retryStrategy: CRTRetryStrategy, allocator: Allocator = defaultAllocator) {
-        let intPointer = UnsafeMutablePointer<Int>.allocate(capacity: 1)
-        intPointer.pointee = 1
-        let atomicVar = aws_atomic_var(value: UnsafeMutableRawPointer(intPointer))
-        let retryStrategyPtr = UnsafeMutablePointer<CRTRetryStrategy>.allocate(capacity: 1)
-        retryStrategyPtr.initialize(to: retryStrategy)
+        let intPointer: UnsafeMutableRawPointer = fromPointer(ptr: 1)
```
Comment (justinboswell) [inline Source/AwsCommonRuntimeKit/io/retryer/CRTAWSRetryToken.swift:10]: Here's that atomic thing again. If Swift doesn't have atomics, we should make a wrapper in this library for a generic atomic<T> where T = integral type or pointer

File: Source/AwsCommonRuntimeKit/mqtt/MqttConnection.swift [hunk 9]
```diff
@@ -462,7 +442,8 @@ public class MqttConnection {
         }, opCallbackPtr)
 
         defer {
-            pointers.deinitializeAndDeallocate()
+            topicPointer.deinitializeAndDeallocate()
```
Comment (justinboswell) [inline Source/AwsCommonRuntimeKit/mqtt/MqttConnection.swift:445]: defer isn't necessary here, topic and payload have already been used

Beginning to feel like we could benefit a ScopedPtr<P> that we can leave on the stack and calls deinitializeAndDeallocate() from its deinit or something

File: Source/AwsCommonRuntimeKit/auth/credentials/CRTCredentialsProvider.swift (hunk 4 above)
Comment (kneekey23) [reply Source/AwsCommonRuntimeKit/auth/credentials/CRTCredentialsProvider.swift:78]: formatted and linted

File: Source/AwsCommonRuntimeKit/io/EventLoopGroup.swift (hunk 6 above)
Comment (kneekey23) [reply Source/AwsCommonRuntimeKit/io/EventLoopGroup.swift:16]: made this!

File: Source/AwsCommonRuntimeKit/auth/credentials/CRTCredentialsProvider.swift (hunk 5 above)
Comment (kneekey23) [reply Source/AwsCommonRuntimeKit/auth/credentials/CRTCredentialsProvider.swift:51]: updated this to use our own Atomic that creates the c type

File: Source/AwsCommonRuntimeKit/mqtt/MqttConnection.swift (hunk 9 above)
Comment (kneekey23) [reply Source/AwsCommonRuntimeKit/mqtt/MqttConnection.swift]: interesting idea, so a pointer as a class that deinitializes itself from its own deinit?

Comment (justinboswell) [reply Source/AwsCommonRuntimeKit/mqtt/MqttConnection.swift]: yes, essentially, it exists on the stack, and when it goes out of scope, it dies. Think of it as an automatically managed temporary

File: Source/AwsCommonRuntimeKit/crt/Atomic.swift [hunk 10]
```diff
@@ -0,0 +1,13 @@
+//  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//  SPDX-License-Identifier: Apache-2.0.
+import AwsCCommon
+
+struct Atomic<I> {
```
Comment (justinboswell) [inline Source/AwsCommonRuntimeKit/crt/Atomic.swift:5]: Hilarious. We don't even access the atomics in Swift. This is something I'll fix later when we move to the FFI library, this is unnecessarily part of the API in use here. Not Swift's problem.

**Review by justinboswell — APPROVED** [review]

File: Source/AwsCommonRuntimeKit/crt/Atomic.swift (hunk 10 above)
Comment (kneekey23) [reply Source/AwsCommonRuntimeKit/crt/Atomic.swift:5]: yep its the c atomic type I need access to so I can pass it but I agree you can abstract that away!

**Review by justinboswell — APPROVED** [review]
